Add select-all / unselect-all to TrackPanel

It's the same as bhofmei/jbplugin-hierarchicalcheckbox#3, because it is related to both repos.

When I tested on `JBrowse 1.12.3`, it works with this [hierarchical check box plugin](https://github.com/bhofmei/jbplugin-hierarchicalcheckbox) , but when I used the `Apollo 2.0.6` (also with `JBrowse 1.12.3` and it's automatically downloaded by `Apollo`. Plugin installation is by specifying setting in `apollo-config.groovy`), it didn't work (no checkboxes show up).

I simply uncomment the line `console.log("plugin HierarchicalCheckboxPlugin constructor")`; in `main.js` of `hierarchical check box plugin` and it shows in console, so the plugin seems to be loaded, but for some reason it didn't work.

Hi @nathandunn , thanks for helping us.

I am working on implementing this features, but I don't know how to configure `javac` options and can't find in the documentation. For example, following are messages when I run `./apollo run-local`. I would like to use such options like `-Xlint:unchecked`, `--stacktrace`, `--info`, and `--debug`. Can you help me to configure these?

```
Installing Perl prerequisites ...Prerequisites installed, finished.
:javac
Note: /Users/hsiaoyi/Documents/workspace/NAL-i5K-Apollo/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:gwtc FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gwtc'.
> Java returned: 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 4 mins 42.268 secs
```

Thank you in advance

--- a/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
+++ b/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
@@ -37,10 +37,12 @@ import org.gwtbootstrap3.client.ui.constants.Toggle;
 import org.gwtbootstrap3.extras.bootbox.client.Bootbox;
 import org.gwtbootstrap3.extras.toggleswitch.client.ui.ToggleSwitch;
 
+import java.awt.Checkbox;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
+import java.util.Locale.Category;
 
 
 /**
@@ -81,6 +83,8 @@ public class TrackPanel extends Composite {
 
     private static Map<String, List<TrackInfo>> categoryMap = new TreeMap<>();
     private static Map<String, Boolean> categoryOpen = new TreeMap<>();
+    private static Map<TrackInfo, CheckBoxButton> checkBoxMap = new TreeMap<>();
+    private static Map<TrackInfo, TrackBodyPanel> trackBodyMap = new TreeMap<>();
 
     public TrackPanel() {
         exportStaticMethod();
@@ -199,6 +203,7 @@ public class TrackPanel extends Composite {
     static class TrackBodyPanel extends PanelBody {
 
         private final TrackInfo trackInfo;
+        
 
         public TrackBodyPanel(TrackInfo trackInfo) {
             this.trackInfo = trackInfo;
@@ -209,7 +214,6 @@ public class TrackPanel extends Composite {
 
             InputGroup inputGroup = new InputGroup();
             addStyleName("track-entry");
-
             final CheckBoxButton selected = new CheckBoxButton();
             selected.setValue(trackInfo.getVisible());
 
@@ -227,17 +231,12 @@ public class TrackPanel extends Composite {
             selected.addValueChangeHandler(new ValueChangeHandler<Boolean>() {
                 @Override
                 public void onValueChange(ValueChangeEvent<Boolean> event) {
-                    JSONObject jsonObject = trackInfo.getPayload();
                     Boolean value = selected.getValue();
-                    trackInfo.setVisible(value);
-                    if (value) {
-                        jsonObject.put("command", new JSONString("show"));
-                    } else {
-                        jsonObject.put("command", new JSONString("hide"));
-                    }
-                    MainPanel.getInstance().postMessage("handleTrackVisibility", jsonObject);
+                    handleSelect(value);
                 }
             });
+
+            checkBoxMap.put(trackInfo,selected);
 
             label.addStyleName("track-link");
             label.setWidth("100%");
@@ -249,11 +248,24 @@ public class TrackPanel extends Composite {
                 }
             }, ClickEvent.getType());
         }
+
+        public void handleSelect(Boolean value) {
+            JSONObject jsonObject = trackInfo.getPayload();
+            trackInfo.setVisible(value);
+            if (value) {
+                jsonObject.put("command", new JSONString("show"));
+            } else {
+                jsonObject.put("command", new JSONString("hide"));
+            }
+            MainPanel.getInstance().postMessage("handleTrackVisibility", jsonObject);
+        }
     }
 
     public void clear() {
         categoryMap.clear();
         categoryOpen.clear();
+        checkBoxMap.clear();
+        trackBodyMap.clear();
         dataGrid.clear();
         dataGrid.add(new org.gwtbootstrap3.client.ui.Label("Loading..."));
     }
@@ -298,13 +310,16 @@ public class TrackPanel extends Composite {
         for (final String key : categoryMap.keySet()) {
             if (key != TrackInfo.TRACK_UNCATEGORIZED) {
 
-                List<TrackInfo> trackInfoList = categoryMap.get(key);
+                final List<TrackInfo> trackInfoList = categoryMap.get(key);
 
                 // if this is a root panel
                 Panel panel = new Panel();
 
                 PanelHeader panelHeader = null;
+                final CheckBoxButton panelSelect = new CheckBoxButton();
+                panelSelect.addStyleName("panel-select");
                 Badge totalBadge = null;
+
                 // if we only have a single uncategorized category, then do not add a header
                 if (categoryOpen.size() != 1 || (!key.equals(TrackInfo.TRACK_UNCATEGORIZED) && categoryOpen.size() == 1)) {
                     panelHeader = new PanelHeader();
@@ -317,6 +332,7 @@ public class TrackPanel extends Composite {
                     totalBadge = new Badge(trackInfoList.size() + "");
                     totalBadge.setPull(Pull.RIGHT);
                     panelHeader.add(totalBadge);
+                    panelHeader.add(panelSelect);
                     panel.add(panelHeader);
                 }
 
@@ -348,8 +364,25 @@ public class TrackPanel extends Composite {
                     for (TrackInfo trackInfo : trackInfoList) {
                         if (trackInfo.getVisible()) ++numVisible;
                         TrackBodyPanel panelBody = new TrackBodyPanel(trackInfo);
+                        trackBodyMap.put(trackInfo, panelBody);
                         panelCollapse.add(panelBody);
                     }
+                    if (numVisible == 0) {
+                        panelSelect.setValue(false);
+                    } else if(numVisible == trackInfoList.size()) {
+                        panelSelect.setValue(true);
+                    }
+                    panelSelect.addValueChangeHandler(new ValueChangeHandler<Boolean>() {
+                        @Override
+                        public void onValueChange(ValueChangeEvent<Boolean> event) {
+                            Boolean value = panelSelect.getValue();
+                            for (TrackInfo trackInfo : trackInfoList) {
+                                checkBoxMap.get(trackInfo).setValue(value);
+                                trackBodyMap.get(trackInfo).handleSelect(value);
+                            }
+                            categoryOpen.put(key, true);
+                        }
+                    });
                 }
                 if (totalBadge != null) {
                     totalBadge.setText(numVisible + "/" + trackInfoList.size());

--- a/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
+++ b/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
@@ -37,12 +37,10 @@ import org.gwtbootstrap3.client.ui.constants.Toggle;
 import org.gwtbootstrap3.extras.bootbox.client.Bootbox;
 import org.gwtbootstrap3.extras.toggleswitch.client.ui.ToggleSwitch;
 
-import java.awt.Checkbox;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
-import java.util.Locale.Category;
 
 
 /**
@@ -318,6 +316,7 @@ public class TrackPanel extends Composite {
                 PanelHeader panelHeader = null;
                 final CheckBoxButton panelSelect = new CheckBoxButton();
                 panelSelect.addStyleName("panel-select");
+                panelSelect.setPull(Pull.RIGHT);
                 Badge totalBadge = null;
 
                 // if we only have a single uncategorized category, then do not add a header
@@ -331,8 +330,8 @@ public class TrackPanel extends Composite {
                     heading.addStyleName("track-header");
                     totalBadge = new Badge(trackInfoList.size() + "");
                     totalBadge.setPull(Pull.RIGHT);
-                    panelHeader.add(totalBadge);
                     panelHeader.add(panelSelect);
+                    panelHeader.add(totalBadge);
                     panel.add(panelHeader);
                 }
 

--- a/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
+++ b/src/gwt/org/bbop/apollo/gwt/client/TrackPanel.java
@@ -37,12 +37,10 @@ import org.gwtbootstrap3.client.ui.constants.Toggle;
 import org.gwtbootstrap3.extras.bootbox.client.Bootbox;
 import org.gwtbootstrap3.extras.toggleswitch.client.ui.ToggleSwitch;
 
-import java.awt.Checkbox;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
-import java.util.Locale.Category;
 
 
 /**

--- a/src/gwt/org/bbop/apollo/gwt/public/theme.css
+++ b/src/gwt/org/bbop/apollo/gwt/public/theme.css
@@ -58,3 +58,8 @@
     text-align: left !important;
     background-color: white !important;
 }
+
+.panel-select {
+    background-color: #f5f5f5 !important; 
+    border: none !important;
+}

